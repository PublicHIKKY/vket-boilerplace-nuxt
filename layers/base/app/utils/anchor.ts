/**
 * @remark aタグを生成して、そのaタグをクリックすることで外部リンクをさせる
 * @param url URL
 * @param blank 新しいタブで開くかどうか(default: true)
 * @param rel rel属性(default: 'norefferer noopener')
 * @param referrerPolicy referrerPolicy属性(default: 'strict-origin-when-cross-origin')
 */
export const makeAnchorElement = (
  url: string,
  blank = true,
  rel = 'norefferer noopener',
  referrerPolicy = 'strict-origin-when-cross-origin',
): HTMLAnchorElement | null => {
  if (!window || !url) return null
  const aElement = document.createElement('a')
  aElement.href = url
  aElement.classList.add('link-via-element-element')
  aElement.referrerPolicy = referrerPolicy
  aElement.rel = rel
  if (blank) aElement.target = '_blank'
  return aElement
}

export const linkViaElement = (
  url: string,
  blank = true,
  rel = 'norefferer noopener',
  referrerPolicy = 'strict-origin-when-cross-origin',
) => {
  const aElement = makeAnchorElement(url, blank, rel, referrerPolicy)
  if (aElement === null) {
    throw new Error('some message')
  }
  aElement.click()
  aElement.remove()
}
