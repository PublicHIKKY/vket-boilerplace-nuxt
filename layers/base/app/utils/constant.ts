/**
 * runtimeConfig・appConfigに置けない・置かない定数を置く場所。
 * 内容がこれなのでユニットテストは必要ない。
 */
export const constant = {
  images: {
    none: '/images/no-image.png',
    none1x1: '/images/no-image_1x1.jpg',
    noneCustom: '/images/no-image-custom.png',
  },
} as const
