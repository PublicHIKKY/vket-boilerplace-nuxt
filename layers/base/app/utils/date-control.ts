import dayjs, { ManipulateType } from 'dayjs'
import customParseFormat from 'dayjs/plugin/customParseFormat'
import isBetween from 'dayjs/plugin/isBetween'
import isSameOrAfter from 'dayjs/plugin/isSameOrAfter'
import isSameOrBefore from 'dayjs/plugin/isSameOrBefore'
import localeData from 'dayjs/plugin/localeData'
import localizedFormat from 'dayjs/plugin/localizedFormat'
import timezone from 'dayjs/plugin/timezone'
import utc from 'dayjs/plugin/utc'

dayjs.locale('ja')

dayjs.extend(localizedFormat)
dayjs.extend(timezone)
dayjs.extend(utc)
dayjs.extend(isBetween)
dayjs.extend(isSameOrBefore)
dayjs.extend(isSameOrAfter)
dayjs.tz.setDefault('Asia/Tokyo')
dayjs.extend(localeData)
dayjs.extend(customParseFormat)

const _locale = {
  0: 'ja',
  1: 'en',
} as const

type LocaleData = (typeof _locale)[keyof typeof _locale]

/**
 * @description 日付取得・フォーマット
 * @returns Result:Date
 */
export function formatDate(
  format: string,
  date: Date | string = new Date(),
): string {
  return dayjs(date).format(format)
}

export function formatEnglishDate(date: Date | string): string {
  return formatDate('LL', date)
}

export function formatJapaneseDate(date: Date | string): string {
  return formatDate('YYYY/MM/DD', date)
}

export function formatEnglishDateTime(date: Date | string): string {
  return formatDate('LL LT', date)
}

export function formatJapaneseDateTime(date: Date | string): string {
  return formatDate('YYYY/MM/DD h:mm A', date)
}

export function formatDateUnixTime(date: Date | string = new Date()): number {
  return dayjs(date).unix()
}

// タイムゾーンをJSTからローカルのタイムゾーンに変更
export function formatJSTtoLocalTimezone(date: string): string {
  const dateJST = dayjs(date).tz('Asia/Tokyo', true)
  const dateUSJ = dayjs(dateJST).tz('UTC').format('YYYY-MM-DDTHH:mm')
  return dayjs(dateUSJ).tz(dayjs.tz.guess()).format('YYYY-MM-DDTHH:mm')
}

// タイムゾーンをローカルのタイムゾーンからJSTに変更
export function formatLocalTimezoneToJST(date: string): string {
  const dateWithTimezone = dayjs(date).tz(dayjs.tz.guess(), true)
  return dayjs(dateWithTimezone).tz('Asia/Tokyo').format('YYYY-MM-DDTHH:mm')
}

export function getCurrentDate(locale: LocaleData = 'ja'): string {
  const date = new Date()
  return locale === 'ja' ? formatJapaneseDate(date) : formatEnglishDate(date)
}

// ユーザのローカルのタイムゾーンを取得
export function getLocalTimezone(): string {
  return 'UTC ' + dayjs().tz(dayjs.tz.guess()).format('Z')
}

// 日付加算されたDateオブジェクトの作成
export function addDateTime(
  value: number,
  unit: ManipulateType = 'day',
  base: Date | string,
) {
  return dayjs(base).add(value, unit).toDate()
}

/**
 * @description 時差対応
 * @returns Result:Date
 */
export function convertTimeToUtc(date: Date | string): string {
  return dayjs(date).utc().format()
}

/**
 * @description 指定日時の比較判定
 * @returns Result:number
 */

export function diffDays(
  to: Date | string,
  from: Date | string | undefined = undefined,
  format: dayjs.UnitType = 'day',
): number {
  const toDate = dayjs(to)
  const fromDate = from ? dayjs(from) : dayjs()
  return fromDate.diff(toDate, format)
}

export function getDiffTimeByUnit(
  to: Date | string,
  from: Date | string | undefined = undefined,
  unitType: dayjs.UnitType = 'day',
): number {
  const millisecond = diffDays(to, from, 'millisecond')
  if (unitType === 'day') {
    return Math.floor(millisecond / 1000 / 60 / 60 / 24)
  }
  if (unitType === 'hour') {
    return Math.floor(millisecond / 1000 / 60 / 60) % 24
  }
  if (unitType === 'minute') {
    return Math.floor(millisecond / 1000 / 60) % 60
  }
  if (unitType === 'second') {
    return Math.floor(millisecond / 1000) % 60
  }
  return -1
}

export function isBeforeTargetDate(date: Date | string): boolean {
  return !!dayjs().isSameOrBefore(dayjs(date))
}

export function isBetweenTargetDates(
  fromDate: Date | string,
  toDate: Date | string,
): boolean {
  return !!dayjs().isBetween(dayjs(fromDate), dayjs(toDate))
}

export function isAfterTargetDate(date: Date | string): boolean {
  return !!dayjs().isSameOrAfter(dayjs(date))
}
