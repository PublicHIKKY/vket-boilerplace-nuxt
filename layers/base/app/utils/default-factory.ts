import exampleRepository from '#base/app/repositories/exampleRepository'
import type { Method } from '#base/app/utils/default-api'

export type Repository = {
  [key in Method]?: { [key: string]: unknown }
}

export type Repositories = {
  [key: string]: Repository
}

export const defaultRepositories = {
  example: exampleRepository,
} as const satisfies Repositories

export type DefaultRepositoryKey = keyof typeof defaultRepositories

export const defaultRepositoryFactory = {
  get: <K extends keyof typeof defaultRepositories>(name: K) =>
    defaultRepositories[name],
}
