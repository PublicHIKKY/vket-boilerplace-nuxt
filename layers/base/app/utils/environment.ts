import { getCurrentInstance } from 'vue'

// 闇魔法
/**
 * setup の中でしか呼べない
 */
export const isNuxtEnvironment = () => !!getCurrentInstance()?.appContext?.app
