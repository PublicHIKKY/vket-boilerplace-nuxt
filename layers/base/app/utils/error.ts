/**
 * ここには到達しない。
 * 適切に switch 文などが書かれているか型レベルでチェックする。
 */
export const unreachable = (_: never): never => {
  throw new Error('unreachable.')
}

/**
 * throw構文を式として使いたい人向けの関数。
 */
export function raiseError(message: string): never {
  throw new Error(message)
}
