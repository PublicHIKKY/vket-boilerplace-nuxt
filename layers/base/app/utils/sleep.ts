import { nextTick } from 'vue'

/**
 * @desc 特定のミリ秒処理を止める。testなどでDOM改変などの非同期に使用
 * @param { number } ms
 */
export const sleep = (ms: number) =>
  new Promise<void>(resolve => setTimeout(() => resolve(), ms))

export const waitEffect = async () => {
  await nextTick()
  await new Promise(resolve => requestAnimationFrame(resolve))
}
