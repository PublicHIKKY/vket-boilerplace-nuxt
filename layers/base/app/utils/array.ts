import { raiseError } from '#base/app/utils/error'

/**
 * `start`から`stop`までの範囲の、数値の配列を生成します。
 *
 * ```ts
 * range(1, 5) // [1, 2, 3, 4, 5]
 * range(1, 5, 2) // [1, 3, 5]
 * range(5, 1, -1) // [5, 4, 3, 2, 1]
 * ```
 */
export const range = (start: number, stop: number, step = 1) =>
  Array.from({ length: (stop - start) / step + 1 }, (_, i) => start + i * step)

/**
 * 配列を逆順にします。
 * `Array.prototype.reverse`は元の配列を書き換えてしまうので、それだと面倒なときに使います。
 *
 * ```ts
 * reversed([1, 2, 3, 4, 5]) // [5, 4, 3, 2, 1]
 * ```
 */
export const reversed = <T>(array: T[]): T[] => [...array].reverse()

/**
 * 配列に値を追加または削除します。
 * 同じ値が入っていた時、両方とも削除します。
 *
 * ```ts
 * toggleList(['a', 'b'], 'a') // ['b']
 * toggleList(['b'], 'a') // ['b', 'a']
 * toggleList(['a', 'b', 'a'], 'a') // ['b']
 * ```
 */
export const toggleList = <T>(list: T[], item: T): T[] =>
  list.includes(item)
    ? list.filter(listItem => listItem !== item)
    : [...list, item]

/**
 * 2つの配列を合体します。
 * 2つの配列の長さが異なる場合、短い方に合わせます。
 *
 * ```typescript
 * zip([1,2,3], [2,3,4,5]) // [[1,2], [2,3], [3,4]]
 * ```
 */
export const zip = <T, U>(xs: T[], ys: U[]): Readonly<[T, U]>[] => {
  const length = xs.length >= ys.length ? ys.length : xs.length
  return range(0, length - 1).map(
    (_, i) =>
      [xs[i] ?? raiseError('Invalid'), ys[i] ?? raiseError('Invalid')] as const,
  )
}

/**
 * JSON.stringify()を利用して、配列が同じ値かを確認します。
 *
 * JSON.stringify()を利用できない値を渡した場合は、例外が出る可能性があります。
 */
export const equal = <T>(xs: T[], ys: T[]): boolean => {
  return JSON.stringify(xs) === JSON.stringify(ys)
}
