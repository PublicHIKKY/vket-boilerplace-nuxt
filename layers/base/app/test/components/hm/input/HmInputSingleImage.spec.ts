import { mount } from '@vue/test-utils'
import HmInputSingleImage from '#base/app/components/hm/input/HmInputSingleImage.vue'
import { waitEffect } from '#base/app/utils/sleep'

beforeEach(() => {
  vi.mock('vue-i18n', () => ({
    useI18n: vi.fn(() => ({
      local: {
        value: 'ja',
      },
      locale: {
        value: 'ja',
      },
      t: (key: string, ..._args: unknown[]) => `dummy-${key}`,
    })),
  }))

  vi.mock('#base/app/utils/file-control', () => ({
    readFileAsBlob: () => 'dummy-blob',
  }))

  URL.createObjectURL = vi.fn(() => 'dummy-for-objectURL')
})

afterEach(() => {
  vi.restoreAllMocks()
})

test('ref component', () => {
  expect(HmInputSingleImage).toBeTruthy()
})

test('mount component', () => {
  const wrapper = mount(HmInputSingleImage, {
    props: {
      defaultImageUrl: null,
    },
  })
  expect(wrapper.getCurrentComponent()).toBeTruthy()
  expect(wrapper.html()).toMatchSnapshot()
})

describe('props', () => {
  test(':optionalAccept', () => {
    const wrapper = mount(HmInputSingleImage, {
      props: {
        optionalAccept: 'image/gif',
        defaultImageUrl: null,
      },
    })
    expect(
      wrapper.find('.hm-single-image-uploader > .input').attributes('accept'),
    ).toMatch('image/gif')
  })

  test(':error', () => {
    const wrapper = mount(HmInputSingleImage, {
      props: {
        error: 'test error',
        defaultImageUrl: null,
      },
    })
    expect(wrapper.find('p[class="error-container"]').text()).toBe('test error')
  })

  test(':isRemovable', () => {
    const wrapper = mount(HmInputSingleImage, {
      props: {
        isRemovable: true,
        defaultImageUrl: 'foo.png',
      },
    })
    expect(wrapper.find('.remove').exists()).toBe(true)
  })

  test(':isRequired', () => {
    const wrapper = mount(HmInputSingleImage, {
      props: {
        isRequired: true,
        defaultImageUrl: null,
      },
    })
    expect(
      wrapper.find('.hm-single-image-uploader > .input').attributes('required'),
    ).toBeDefined()
  })

  test(':needCropper, :cropWidth, and :cropHeight', async () => {
    const wrapper = mount(HmInputSingleImage, {
      props: {
        needCropper: true,
        cropWidth: undefined,
        cropHeight: undefined,
        defaultImageUrl: null,
      },
    })

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    await (wrapper as any).vm.changeImage([
      new File([], 'foo.png'),
    ] as any as FileList) // eslint-disable-line @typescript-eslint/no-explicit-any
    await waitEffect()

    expect(wrapper.find('.ha-dialog').exists()).toBe(true)
  })
})
