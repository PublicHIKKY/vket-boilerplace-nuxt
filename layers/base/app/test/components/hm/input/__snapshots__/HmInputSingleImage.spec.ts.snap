// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mount component 1`] = `
"<div data-v-c9c71ad2="" class="hm-input-single-image">
  <div data-v-c9c71ad2="" class="wrapper">
    <!-- TODO: エラーメッセージの表示をする際に、必要に応じてHmInputTextBase.vue同様の修正(DOM構造とエラーmsgのstyle)を行う --><label data-v-8323f5e3="" data-v-c9c71ad2="" class="hm-input-file hm-single-image-uploader"><input data-v-d0a26e0a="" data-v-8323f5e3="" class="ha-base-input input" type="file" accept="image/png,image/jpeg" name="file" required="">
      <div data-v-8323f5e3="" class="image-box">
        <div data-v-8323f5e3="" class="inner"><span data-v-8323f5e3="" class="text"> Select file or drag it! </span></div>
      </div>
    </label>
  </div>
  <p data-v-c9c71ad2="" class="error-container">
    <!--v-if-->
  </p>
  <!--v-if-->
  <!--v-if-->
</div>"
`;
